### PR TITLE
Add index_rows and index_cols

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by konradjk** [comment]
Docs and interface looks good to me!

**Comment by konradjk** [comment]
What are the names gonna be now? I kinda liked `number_rows` (after all the back and forth we had)

**Comment by tpoterba** [comment]
am no working on changing to the following:

these are renamed to index_rows, index_cols
the existing index_X methods are renamed to view_join_X
Table.indexed => Table.index
Table.range produces a field 'idx' not 'index'

**Comment by konradjk** [comment]
Ok, I like all but `view_join_X` - I hate it less than what it was before (i.e. `index_X`) but it's not super intuitive

**Comment by tpoterba** [comment]
yeah, still deciding if that should even be exposed.

**Comment by konradjk** [comment]
I think you need it, no? I thought there was at least one join that was impossible without it

**Comment by tpoterba** [comment]
product joins, yeah. But we could solve that with some sort of group_by_key method

File: python/hail/api2/matrixtable.py [hunk 1]
```diff
@@ -2210,6 +2210,64 @@ def unpersist(self):
...
+        -----
+        The field added is type :class:`.TInt64`.
+
+        The row index is 0-indexed; the values are found in the range
+        ``[0, N)``.
+
+        Parameters
+        ----------
+        name : :obj:`str`
+            Name for row number field.
```
Comment (jbloom22) [inline python/hail/api2/matrixtable.py]: number => index

File: python/hail/api2/matrixtable.py [hunk 2]
```diff
@@ -2210,6 +2210,64 @@ def unpersist(self):
...
+        -----
+        The field added is type :class:`.TInt32`.
+
+        The column index is 0-indexed; the values are found in the range
+        ``[0, N)``.
+
+        Parameters
+        ----------
+        name: :obj:`str`
+            Name for column number field.
```
Comment (jbloom22) [inline python/hail/api2/matrixtable.py]: number => index

File: python/hail/api2/matrixtable.py [hunk 3]
```diff
@@ -2210,6 +2210,64 @@ def unpersist(self):
         """
         return MatrixTable(self._jvds.unpersist())
 
+    @handle_py4j
+    @typecheck_method(name=strlike)
+    def index_rows(self, name='row_idx'):
+        """Add the row integer index as a new field.
```
Comment (jbloom22) [inline python/hail/api2/matrixtable.py]: I prefer "Add the integer index of each row as a new row field." to "row integer index".

Currently index on table says: "Add the numerical index of each row as a new field." and "The value of this column is the numerical index of each row".  Can you change numerical to integer there?

File: python/hail/api2/matrixtable.py [hunk 4]
```diff
@@ -2210,6 +2210,64 @@ def unpersist(self):
...
+        -------
+        :class:`.MatrixTable`
+            Dataset with new field.
+        """
+        return MatrixTable(self._jvds.indexRows(name))
+
+    @handle_py4j
+    @typecheck_method(name=strlike)
+    def index_cols(self, name='col_idx'):
+        """Add the column integer index as a new field.
```
Comment (jbloom22) [inline python/hail/api2/matrixtable.py]: "Add the integer index of each column as a new column field."

File: python/hail/api2/table.py [hunk 5]
```diff
@@ -1315,7 +1315,7 @@ def format_type(typ):
     @handle_py4j
     @typecheck_method(name=strlike)
     def index(self, name='idx'):
-        """Add the numerical index of each row as a new field.
+        """Add the integer index of each row as a new row field.
```
Comment (jbloom22) [inline python/hail/api2/table.py]: change `numerical` to `integer` below as well

File: python/hail/api2/table.py [hunk 6]
```diff
@@ -1162,7 +1162,7 @@ def range(cls, n, num_partitions=None):
         -----
         The resulting table has one column:
 
-         - `idx` (**Int32**) - Unique row index from 0 to ``N - 1``
+         - `idx` (**Int32**) - Unique row index from 0 to `n`.
```
Comment (jbloom22) [inline python/hail/api2/table.py]: This should still be `n - 1`

File: src/main/scala/is/hail/variant/MatrixTable.scala [hunk 7]
```diff
@@ -2699,15 +2699,15 @@ g = let newgt = gtIndex(oldToNew[gtj(g.GT)], oldToNew[gtk(g.GT)]) and
     val newVAType = newRowType.asInstanceOf[TStruct].fieldType(2)
     val localRowType = rowType
 
-    val partitionStarts = partitionCounts().scanLeft(0L)(_ + _)
+    val starts = partitionStarts()
```
Comment (jbloom22) [inline src/main/scala/is/hail/variant/MatrixTable.scala]: we use `partStarts` elsewhere for the non-function

**Pull request merged**